### PR TITLE
[Merged by Bors] - feat: norm num extension for division and succ of naturals

### PR DESCRIPTION
**Pull request by alexjbest**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by fpvandoorn** [comment]
LGTM

could you add a few tests?

Looking at `lean4/src/runtime/object.h` there one more Nat operation that the kernel recognizes and we don't have a norm-num extension for, which is `Nat.succ`.

bors d+

File: Mathlib/Tactic/NormNum/Basic.lean [hunk 1]
```diff
@@ -927,3 +927,22 @@ such that `norm_num` successfully recognises both `a` and `b`. -/
...
   let r : Q(Nat.mod $na $nb = $nc) := (q(Eq.refl $nc) : Expr)
   return (.isNat sℕ nc q(isNat_natMod $pa $pb $r) : Result q($a % $b))
+
+theorem isNat_natDiv : {a b : ℕ} → {a' b' c : ℕ} →
+    IsNat a a' → IsNat b b' → Nat.div a' b' = c → IsNat (a / b) c
+  | _, _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, rfl => ⟨by aesop⟩
+
+/-- The `norm_num` extension which identifies expressions of the form `Nat.div a b`,
+such that `norm_num` successfully recognises both `a` and `b`. -/
+@[norm_num (_ : ℕ) % _, HDiv.hDiv (_ : ℕ) _, Nat.div _ _] def evalNatDiv :
```
Comment (digama0) [inline Mathlib/Tactic/NormNum/Basic.lean:937]: I think the first two patterns here are the same. You probably want `Div.div` for the second pattern

Comment (alexjbest) [reply Mathlib/Tactic/NormNum/Basic.lean]: oops, looks like the first pattern was wrong even

File: Mathlib/Tactic/NormNum/Basic.lean [hunk 2]
```diff
@@ -927,3 +927,22 @@ such that `norm_num` successfully recognises both `a` and `b`. -/
...
   let r : Q(Nat.mod $na $nb = $nc) := (q(Eq.refl $nc) : Expr)
   return (.isNat sℕ nc q(isNat_natMod $pa $pb $r) : Result q($a % $b))
+
+theorem isNat_natDiv : {a b : ℕ} → {a' b' c : ℕ} →
+    IsNat a a' → IsNat b b' → Nat.div a' b' = c → IsNat (a / b) c
+  | _, _, _, _, _, ⟨rfl⟩, ⟨rfl⟩, rfl => ⟨by aesop⟩
+
+/-- The `norm_num` extension which identifies expressions of the form `Nat.div a b`,
+such that `norm_num` successfully recognises both `a` and `b`. -/
+@[norm_num (_ : ℕ) / _, Div.Div (_ : ℕ) _, Nat.div _ _] def evalNatDiv :
```
Comment (alexjbest) [inline Mathlib/Tactic/NormNum/Basic.lean:937]:
```suggestion
@[norm_num (_ : ℕ) / _, Div.div (_ : ℕ) _, Nat.div _ _] def evalNatDiv :
```

**Comment by alexjbest** [comment]
I added Nat.succ support so maybe @digama0 or @fpvandoorn wants to take another look?

**Comment by digama0** [comment]
Regarding `Nat.succ`, I think we want to have special support to force a sequence of `succ` applied to `zero` or a numeral to a numeral in one go via `rfl`, rather than doing it one application at a time.

**Comment by alexjbest** [comment]
> Regarding `Nat.succ`, I think we want to have special support to force a sequence of `succ` applied to `zero` or a numeral to a numeral in one go via `rfl`, rather than doing it one application at a time.

Do you want to push what you have in mind? Or I can revert the succ bit from this PR. (Or merge this and improve later)
I'm not sure I can/should add this more advanced approach to this PR.

**Comment by digama0** [comment]
LGTM. I'll do it in a follow up PR.

bors r+

**Pull request closed without merging**